File: pkg_rospy/scripts/simple_action_server.py
#! /usr/bin/env python

#Counts until count_upto with wait_time between each count and sends the result on completion

import rospy
import actionlib
from pkg_rospy.msg import CountNumbersAction, CountNumbersGoal, CountNumbersResult

class CountNumbersServer:
    def __init__(self):
        self._as = actionlib.SimpleActionServer('/count_numbers', CountNumbersAction, 
                execute_cb = self.on_goal, auto_start=False)
        self._as.start()
        rospy.loginfo("count numbers server started")

        self._counter = 0
    
    def on_goal(self, goal):
        rospy.loginfo("Goal has be en received")
        rospy.loginfo(goal)
        count_upto = goal.count_upto
        wait_time = goal.wait_time

        self._counter = 0
        rate = rospy.Rate(1.0/wait_time)

        while self._counter < count_upto:
            self._counter += 1
            rate.sleep()
        
        result = CountNumbersResult()
        result.final_count = self._counter
        self._as.set_succeeded(result)
        rospy.loginfo("result has been sent")

def main():
    rospy.init_node("count_numbers_server")
    Server = CountNumbersServer()
    rospy.spin()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: pkg_roscpp/src/simple_action_server.cpp
#include "ros/ros.h"
#include "actionlib/server/simple_action_server.h"
#include "pkg_roscpp/CountNumbersAction.h"

class ActionServer{
protected:
    ros::NodeHandle nh_;
    actionlib::SimpleActionServer<pkg_roscpp::CountNumbersAction> as_;

    pkg_roscpp::CountNumbersResult result_;    

public:
    ActionServer(std::string name) : 
        as_(nh_, name, boost::bind(&ActionServer::executeCB, this, _1), false)
    {
        as_.start();
        ROS_INFO("Action Server started");
    }

    ~ActionServer(void){}

    void executeCB(const pkg_roscpp::CountNumbersGoalConstPtr &goal){
        int32_t count = 0;
        int32_t count_upto = goal->count_upto;
        float wait_time = goal->wait_time;
        ROS_INFO("Goal received, count_upto: [%d], wait_time: [%f]", count_upto, wait_time);

        ros::Rate rate(1.0/wait_time);

        while (!ros::isShuttingDown() && count<count_upto){
            count++;
            rate.sleep();
        }
        result_.final_count = count;
        as_.setSucceeded(result_);
        ROS_INFO("Goal sent");
    }
};

int main(int argc, char **argv){
    ros::init(argc, argv, "simple_action_server");
    ActionServer Server("count_numbers");

    ros::spin();
    
    return 0;
}